### PR TITLE
*: add context for preprocess

### LINKED ISSUE #38359: Add context (`context.Context`) for planner preprocess
## Enhancement

The planner preprocess procedure doesn't pass in the `ctx context.Context`. However, it's needed by https://github.com/pingcap/tidb/issues/38274, as the stale-read preprocessor should send request to the pd with this ctx. The context is needed for killing the connection (and global deadline, tracing...).

It will bring a lot of changes to the related functions / tests. I don't know whether there are better ways to handle this.

### PR DESCRIPTION
**Pull request by YangKeao**
Signed-off-by: YangKeao <yangkeao@chunibyo.icu>

### What problem does this PR solve?

Issue Number: close #38359

Problem Summary:

The `Preprocess` function doesn't receive and pass in the `ctx context.Context` arguments. As described in #38359, it's needed by the external-timestamp read feature.

### What is changed and how it works?

Add `ctx context.Context` arguments for every related functions. And I don't change the ast visitor interface (because it will cause too many changes), but add a field ctx in in the preprocessor struct.

### CODE REVIEW

**Review by xhebox — APPROVED** [review]

**Review by bb7133 — APPROVED** [review]
LGTM

File: sessiontxn/staleread/processor.go [hunk 1]
```diff
@@ -59,7 +60,8 @@ type baseProcessor struct {
 	is          infoschema.InfoSchema
 }
 
-func (p *baseProcessor) init(sctx sessionctx.Context) {
+func (p *baseProcessor) init(ctx context.Context, sctx sessionctx.Context) {
+	p.ctx = ctx
```
Comment (lcwangchao) [inline sessiontxn/staleread/processor.go:64]: If processor has the `context.Context`, can we replace `context.TODO()` with `p.ctx` in this file?

Comment (YangKeao) [reply sessiontxn/staleread/processor.go:64]: Sure! Replaced :handshake:

**Review by hawkingrei — APPROVED** [review]

**Comment by YangKeao** [comment]
@hawkingrei @xhebox Could you help me to merge this PR?

**Pull request merged**